go/libraries/doltcore/schema/encoding/serialization.go

```diff
@@ -0,0 +1,472 @@
+		for j := len(tags) - 1; j >= 0; j-- {
+			pos := ordinalMap[tags[j]]
+			b.PrependUint16(uint16(pos))
+		}
+		ico := b.EndVector(len(tags))
+
+		// serialize key columns
+		tags = idx.AllTags()
+		serial.IndexStartKeyColumnsVector(b, len(tags))
+		for j := len(tags) - 1; j >= 0; j-- {
+			pos := ordinalMap[tags[j]]
+			b.PrependUint16(uint16(pos))
+		}
+		ko := b.EndVector(len(tags))
+
+		serial.IndexStart(b)
+		serial.IndexAddName(b, no)
+		serial.IndexAddComment(b, co)
+		serial.IndexAddIndexColumns(b, ico)
+		serial.IndexAddKeyColumns(b, ko)
+		serial.IndexAddPrimaryKey(b, false)
+		serial.IndexAddUniqueKey(b, idx.IsUnique())
+		serial.IndexAddSystemDefined(b, !idx.IsUserDefined())
+		offs[i] = serial.IndexEnd(b)
+	}
+
+	serial.TableSchemaStartSecondaryIndexesVector(b, len(indexes))
+	for i := len(offs) - 1; i >= 0; i-- {
+		b.PrependUOffsetT(offs[i])
+	}
+	return b.EndVector(len(indexes))
+}
+
+func deserializeSecondaryIndexes(sch schema.Schema, s *serial.TableSchema) error {
+	idx := new(serial.Index)
+	col := new(serial.Column)
+	for i := 0; i < s.SecondaryIndexesLength(); i++ {
+		s.SecondaryIndexes(idx, i)
+		assertTrue(!idx.PrimaryKey())
+
+		name := string(idx.Name())
+		props := schema.IndexProperties{
+			IsUnique:      idx.UniqueKey(),
+			IsUserDefined: !idx.SystemDefined(),
+			Comment:       string(idx.Comment()),
+		}
+
+		tags := make([]uint64, idx.IndexColumnsLength())
+		for j := range tags {
+			pos := idx.IndexColumns(j)
+			s.Columns(col, int(pos))
+			tags[j] = col.Tag()
+		}
+
+		_, err := sch.Indexes().AddIndexByColTags(name, tags, props)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func serializeChecks(b *fb.Builder, checks []schema.Check) fb.UOffsetT {
+	offs := make([]fb.UOffsetT, len(checks))
+	for i := len(offs) - 1; i >= 0; i-- {
+		eo := b.CreateString(checks[i].Expression())
+		no := b.CreateString(checks[i].Name())
+		serial.CheckConstraintStart(b)
+		serial.CheckConstraintAddEnforced(b, checks[i].Enforced())
+		serial.CheckConstraintAddExpression(b, eo)
+		serial.CheckConstraintAddName(b, no)
+		offs[i] = serial.CheckConstraintEnd(b)
+	}
+
+	serial.TableSchemaStartChecksVector(b, len(checks))
+	for i := len(offs) - 1; i >= 0; i-- {
+		b.PrependUOffsetT(offs[i])
+	}
+	return b.EndVector(len(checks))
+}
+
+func deserializeChecks(sch schema.Schema, s *serial.TableSchema) error {
+	coll := sch.Checks()
+	c := new(serial.CheckConstraint)
+	for i := 0; i < s.ChecksLength(); i++ {
+		s.Checks(c, i)
+		n, e := string(c.Name()), string(c.Expression())
+		if _, err := coll.AddCheck(n, e, c.Enforced()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func keylessSerialSchema(s *serial.TableSchema) bool {
+	n := s.ColumnsLength()
+	if n < 2 {
+		return false
+	}
+	// keyless id is the 2nd to last column
+	// in the columns vector (by convention)
+	// and the only field in key tuples of
+	// the clustered index.
+	id := serial.Column{}
+	s.Columns(&id, n-2)
```

Comment:
Nit: Is this access pattern better than `id := new(serial.Column)` for heap allocation / elision?

If so, should we be using it above in the functions like `deserializeSecondaryIndexes`?